From 7a4f4746713942d1acc7b9dd2b01f0961fb805c0 Mon Sep 17 00:00:00 2001
From: Joel Hockey <joelhockey@chromium.org>
Date: Tue, 5 Feb 2019 13:38:06 -0800
Subject: [PATCH 8/9] shared: Progress metadata as a map

Modify the progress metadata to send stage, percent, speed in
a map keyed by 'progress'. Metadata format will now be:
* progress => {stage: <string>, percent: <string>, speed: <string>}
* <stage>_progress => <string>

Signed-off-by: Joel Hockey <joelhockey@chromium.org>
---
 lxd/containers_post.go |  2 +-
 lxd/images.go          |  4 ++--
 shared/util.go         | 10 ++++++----
 3 files changed, 9 insertions(+), 7 deletions(-)

diff --git a/lxd/containers_post.go b/lxd/containers_post.go
index e6d8140a..dc98b9b0 100644
--- a/lxd/containers_post.go
+++ b/lxd/containers_post.go
@@ -122,7 +122,7 @@ func createFromImage(d *Daemon, req *api.ContainersPost) Response {
 			return err
 		}
 
-		metadata := make(map[string]string)
+		metadata := make(map[string]interface{})
 		_, err = containerCreateFromImage(d, args, info.Fingerprint, &ioprogress.ProgressTracker{
 			Handler: func(percent, speed int64) {
 				shared.SetProgressMetadata(metadata, "create_container_from_image_unpack", "Unpack", percent, speed)
diff --git a/lxd/images.go b/lxd/images.go
index 68110fb1..fb5af933 100644
--- a/lxd/images.go
+++ b/lxd/images.go
@@ -206,7 +206,7 @@ func imgPostContInfo(d *Daemon, r *http.Request, req api.ImagesPost, op *operati
 	}
 
 	// Track progress writing tarfile
-	metadata := make(map[string]string)
+	metadata := make(map[string]interface{})
 	tarfileProgressWriter := &ioprogress.ProgressWriter{
 		WriteCloser: tarfile,
 		Tracker: &ioprogress.ProgressTracker{
@@ -261,7 +261,7 @@ func imgPostContInfo(d *Daemon, r *http.Request, req api.ImagesPost, op *operati
 		}
 
 		// Track progress writing gzipped file
-		metadata = make(map[string]string)
+		metadata = make(map[string]interface{})
 		tarfileProgressReader := &ioprogress.ProgressReader{
 			ReadCloser: tarfile,
 			Tracker: &ioprogress.ProgressTracker{
diff --git a/shared/util.go b/shared/util.go
index caf34088..199e78fc 100644
--- a/shared/util.go
+++ b/shared/util.go
@@ -1016,11 +1016,13 @@ func DownloadFileSha512(httpClient *http.Client, useragent string, progress func
 	return DownloadFileHash(httpClient, useragent, progress, canceler, filename, url, hash, sha512.New(), target)
 }
 
-func SetProgressMetadata(metadata map[string]string, stage, displayPrefix string, percent, speed int64) {
+func SetProgressMetadata(metadata map[string]interface{}, stage, displayPrefix string, percent, speed int64) {
+	progress := make(map[string]string)
 	// stage, percent, speed sent for API callers.
-	metadata["progress_stage"] = stage
-	metadata["progress_percent"] = strconv.FormatInt(percent, 10)
-	metadata["progress_speed"] = strconv.FormatInt(speed, 10)
+	progress["stage"] = stage
+	progress["percent"] = strconv.FormatInt(percent, 10)
+	progress["speed"] = strconv.FormatInt(speed, 10)
+	metadata["progress"] = progress
 	// <stage>_progress with formatted text sent for lxc cli.
 	metadata[stage+"_progress"] = fmt.Sprintf("%s: %d%% (%s/s)", displayPrefix, percent, GetByteSizeString(speed, 2))
 }
-- 
2.20.1.791.gb4d0f1c61a-goog

